src/Data/Singletons/Promote.hs

```diff
@@ -816,6 +924,69 @@ promoteLetDecRHS rhs_sort type_env fix_env mb_let_uniq name let_dec_rhs = do
     count_args (DClause pats _ : _) = return $ length pats
     count_args _ = fail $ "Impossible! A function without clauses."
 
+-- An auxiliary data type used in promoteLetDecRHS that describes information
+-- related to the promoted kind of a class method default or normal
+-- let binding.
+data LetDecRHSKindInfo =
+  LDRKI (Maybe DKind) -- The standalone kind signature, if applicable.
+                      -- This will be Nothing if the let-dec RHS has local
+                      -- variables that it closes over.
+                      -- See Note [No SAKs for let-decs with local variables]
+        [DTyVarBndr]  -- The type variable binders of the kind.
+        [DKind]       -- The argument kinds.
+        DKind         -- The result kind.
+
+{-
+Note [No SAKs for let-decs with local variables]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Consider promoting this:
+
+  f :: Bool
+  f = let x = True
+          g :: () -> Bool
+          g _ = x
+      in g ()
+
+Clearly, the promoted `F` type family will have the following SAK:
+
+  type F :: ()
+
+What about `G`? At a passing glance, it appears that you could get away with
+this:
+
+  type G :: Bool -> ()
+
+But this isn't quite right, since `g` closes over `x = True`. The body of `G`,
+therefore, has to lift `x` to be an explicit argument:
+
+  type family G x (u :: ()) :: Bool where
+    G x _ = x
+
+At present, we don't keep track of the types of local variables like `x`, which
+makes it difficult to create a SAK for things like `G`. Here are some possible
+ideas, each followed by explanations for why they are infeasible:
+
+* Use wildcards:
+
+    type G :: _ -> () -> Bool
+
+  Alas, GHC currently does not allow wildcards in SAKs. See GHC#17432.
+
+* Use visible dependent quantification to avoid having to say what the kind
+  of `x` is:
+
+    type G :: forall x -> () -> Bool
+
+  A clever trick to be sure, but it doesn't quite do what we want, since
+  GHC will generalize that kind to become `forall (x :: k) -> () -> Bool`,
```

Comment:
It's worse than that, you'd also run into #17072

Reply:
That's true, although GHC#17072 is unlikely to be an issue in this particular circumstance. The promoted versions of local functions are, by virtue of being local, an internal implementation detail, so we are free to choose whatever kinds we want for them. There may be other problems with using SAKS here besides excessive kind generalization that I haven't thought of, but I would be surprised if GHC#17072 were one of them.